File: src/main/java/voidchess/player/ki/ComputerPlayer.java
package voidchess.player.ki;

import voidchess.board.ChessGameInterface;
import voidchess.helper.Move;
import voidchess.helper.PawnPromotion;
import voidchess.helper.Position;
import voidchess.helper.RuntimeFacade;
import voidchess.player.PlayerInterface;
import voidchess.player.ki.concurrent.ConcurrencyStrategy;
import voidchess.player.ki.concurrent.ConcurrencyStrategyFactory;
import voidchess.player.ki.evaluation.Evaluated;
import voidchess.player.ki.evaluation.EvaluatedMove;
import voidchess.player.ki.openings.OpeningsLibrary;
import voidchess.ui.ComputerPlayerUI;
import voidchess.ui.TableInterface;

import java.util.Iterator;
import java.util.List;
import java.util.NavigableSet;
import java.util.Random;

/**
 * @author stephan
 */
public class ComputerPlayer
        implements PlayerInterface {
    private ChessGameInterface game;
    private TableInterface table;
    private ComputerPlayerUI ui;
    private DynamicEvaluation dynamicEvaluation;
    private SearchTreePruner standardPruner;
    private StaticEvaluationInterface standardEvaluation;
    private ConcurrencyStrategy concurrencyStrategy;
    private boolean usesStandardEvaluation;
    private OpeningsLibrary openingsLibrary;
    private boolean useLibrary;
    private Random randomNumberGenerator;

    public ComputerPlayer(TableInterface table, ChessGameInterface game, ComputerPlayerUI ui) {
        this.ui = ui;
        this.game = game;
        this.table = table;

        standardPruner = new SimplePruner(1, 2, 2);
        standardEvaluation = new StaticEvaluation();
        dynamicEvaluation = new DynamicEvaluation(standardPruner, standardEvaluation);
        concurrencyStrategy = ConcurrencyStrategyFactory.getConcurrencyStrategy(ui, 1);
        openingsLibrary = new OpeningsLibrary("openings.txt");
        randomNumberGenerator = new Random();
        initEvaluation();
    }

    public void play() {

        EvaluatedMove chosenMove = getNextMove();
        ui.setValue(chosenMove.getValue(), chosenMove.getMove());
        table.move(chosenMove.getMove());

        RuntimeFacade.collectGarbage();
    }

    private EvaluatedMove getNextMove() {
        //lets see if the library contains a next move
        EvaluatedMove nextMove = lookUpNextMove();
        //we can always compute a next move
        if (nextMove == null) {
            nextMove = computeNextMove();
        }
        return nextMove;
    }

    private EvaluatedMove lookUpNextMove() {
        if (useLibrary) {
            String history = game.getCompleteHistory();
            List<Move> possibleMoves = openingsLibrary.nextMove(history);
            if (!possibleMoves.isEmpty()) {
                //display that the computer is working
                ui.setProgress(0, 1);
                //pick a random move
                Move randomMove = possibleMoves.get(randomNumberGenerator.nextInt(possibleMoves.size()));
                //and evaluate it
                boolean isWhitePlayer = game.isWhiteTurn();
                game.move(randomMove);
                Evaluated evaluation = standardEvaluation.getPrimaryEvaluation(game, isWhitePlayer);
                standardEvaluation.addSecondaryEvaluation(game, isWhitePlayer, evaluation);
                game.undo();

                //wait before playing so that the user can clearly see the computer's move
                wait(300);

                return new EvaluatedMove(randomMove, evaluation);
            }
        }
        //the library has no more information on this sequence
        useLibrary = false;
        return null;
    }

    private EvaluatedMove computeNextMove() {
        //the game starts with a static game evaluation that considers many things
        //but in the endgame, when only the king of one side can move than the only important thing is
        //how many possible moves the king has left. So the Evaluation strategy has to be changed.
        pickStaticSpaceEvaluationIfNeccessary();

//      TODO  real Timer
//		long time = System.currentTimeMillis();

        NavigableSet<EvaluatedMove> sortedEvaluatedMoves = concurrencyStrategy.evaluatePossibleMoves(game, dynamicEvaluation);

//		final int calls = StaticEvaluation.getCallCounter();
//		final int totalNumberOfMoves = evaluation.totalNumberOfMoves();
//		long duration = System.currentTimeMillis()-time;
//		double timePerMove = duration/(double)totalNumberOfMoves;
//		double timePerCall = duration/(double)calls;
//		System.out.println( "ComputerPlayer.play()" );
//		System.out.println( "moves:       "+totalNumberOfMoves );
//		System.out.println( "staticEval:  "+calls );
//		System.out.println( "duration:    "+(duration)+"ms" );
//		System.out.println( "timePerMove: "+(timePerMove)+"ms" );
//		System.out.println( "timePerCall: "+(timePerCall)+"ms" );
//		System.out.println();

        EvaluatedMove chosenMove = pickNextMoveByEvaluation(sortedEvaluatedMoves);
        return chosenMove;
    }

    /**
     * pick the move to make given an evaluated Set of possible moves.
     * (Taking always the best move would lead to a limited number of openings!)
     *
     * @param sortedEvaluatedMoves (set.first is the best move for the ki, set.last the worst)
     * @return the move the ki will make next
     */
    private EvaluatedMove pickNextMoveByEvaluation(NavigableSet<EvaluatedMove> sortedEvaluatedMoves) {
        Iterator<EvaluatedMove> evaluation = sortedEvaluatedMoves.descendingIterator();
        EvaluatedMove bestMove = evaluation.next();

        //as long as the top moves are almost equally good, pick randomly one (with a higher chance for the better move)
        EvaluatedMove chosenMove = bestMove;
        while (evaluation.hasNext()) {
            if (Math.random() < 0.6) break;
            EvaluatedMove tempMove = evaluation.next();
            if (tempMove.getValue().isCloseToByCombined(bestMove.getValue())) {
                chosenMove = tempMove;
            } else {
                break;
            }
        }
        return chosenMove;
    }

    /**
     * the game starts with a static game evaluation that considers many things
     * but in the endgame, when only the king of one side can move than the
     * only important thing is how many possible moves the king has left.
     * So the Evaluation strategy has to be changed.
     */
    private void pickStaticSpaceEvaluationIfNeccessary() {
        if (usesStandardEvaluation && StaticSpaceEvaluation.shouldUseStaticSpaceEvaluation(game)) {
            //wird genau einmal pro Spiel ausgeführt
            dynamicEvaluation.setEvaluationStrategy(new StaticSpaceEvaluation());
            dynamicEvaluation.setSearchTreePruner(new FullMovePruner(2, 2, 2));
            usesStandardEvaluation = false;
        }
    }

    public PawnPromotion askForPawnPromotionType(Position pawnPosition) {
        return PawnPromotion.QUEEN;
    }

    //am Anfang jedes Spiels wird die EvaluationStrategie zurückgesetzt vom potentiel StaticSpaceEvaluation
    public void setIsPlaying(boolean isPlaying) {
        if (isPlaying) initEvaluation();
    }

    //setzt die Default-EvaluationStrategie
    private void initEvaluation() {
        dynamicEvaluation.setEvaluationStrategy(standardEvaluation);
        dynamicEvaluation.setSearchTreePruner(standardPruner);
        usesStandardEvaluation = true;
        //use the library only if the figures are used in the classical way (no Chess960)
        useLibrary = game.isStandardGame();
    }

    //die Funktion wird nur von HumanPlayer benötigt (bis jetzt)
    public void setColor(boolean isWhite) {
    }

    public void setSearchTreePruner(SearchTreePruner pruner) {
        standardPruner = pruner;
        dynamicEvaluation.setSearchTreePruner(pruner);
    }

    public void setNumberOfCoresToUse(int numberOfCoresToUse) {
        concurrencyStrategy = ConcurrencyStrategyFactory.getConcurrencyStrategy(ui, numberOfCoresToUse);
    }

    private void wait(final int milliseconds) {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException e) {
        }
    }
}

File: src/main/kotlin/voidchess/player/ki/ComputerPlayer.kt
package voidchess.player.ki

import voidchess.board.ChessGameInterface
import voidchess.board.move.Move
import voidchess.board.move.MoveResult
import voidchess.board.move.PawnPromotion
import voidchess.board.move.Position
import voidchess.helper.RuntimeFacade
import voidchess.player.PlayerInterface
import voidchess.player.ki.concurrent.ConcurrencyStrategy
import voidchess.player.ki.concurrent.getConcurrencyStrategy
import voidchess.player.ki.evaluation.*
import voidchess.player.ki.openings.OpeningsLibrary
import voidchess.ui.ComputerPlayerUI
import voidchess.ui.TableInterface
import voidchess.ui.Thumb
import java.text.DecimalFormat
import java.util.*


class ComputerPlayer(private val table: TableInterface, private val game: ChessGameInterface, private val ui: ComputerPlayerUI) : PlayerInterface {
    private val evaluatingMinMax: EvaluatingMinMax
    private var standardPruner: SearchTreePruner
    private val standardEvaluation: EvaluatingStatically
    private var concurrencyStrategy: ConcurrencyStrategy
    private var usesStandardEvaluation: Boolean = false
    private val openingsLibrary: OpeningsLibrary
    private var useLibrary: Boolean = false
    private val randomNumberGenerator: Random
    private var isWhite = false

    init {
        standardPruner = PrunerWithIrreversibleMoves(1, 2, 2)
        standardEvaluation = EvaluatingAsIsNow
        evaluatingMinMax = EvaluatingMinMax(standardPruner, standardEvaluation)
        concurrencyStrategy = getConcurrencyStrategy(ui::setProgress, 1)
        openingsLibrary = OpeningsLibrary("openings.txt")
        randomNumberGenerator = Random()
        reset()
    }

    override fun play() {
        ui.setBubbleText(null)
        ui.showThoughts(true)
        val chosenMove = nextMove()
        ui.setValue(chosenMove.value)
        table.move(chosenMove.move)

        RuntimeFacade.collectGarbage()
    }

    //lets see if the library contains a next move, else we compute the next move
    private fun nextMove() = lookUpNextMove() ?: computeNextMove()

    private fun lookUpNextMove(): EvaluatedMove? {
        fun wait(milliseconds: Int) = try {Thread.sleep(milliseconds.toLong())} catch (e: InterruptedException) {}

        if (useLibrary) {
            val history = game.getCompleteHistory()
            val possibleMoves = openingsLibrary.nextMove(history)
            if (possibleMoves.isNotEmpty()) {
                //display that the computer is working
                ui.setProgress(0, 1)
                //pick a random move
                val randomMove = possibleMoves[randomNumberGenerator.nextInt(possibleMoves.size)]
                //and evaluate it
                val isWhitePlayer = game.isWhiteTurn
                game.move(randomMove)
                val preliminaryEvaluation = standardEvaluation.getPreliminaryEvaluation(game, isWhitePlayer)
                val fullEvaluation = preliminaryEvaluation + standardEvaluation.getSecondaryEvaluation(game, isWhitePlayer)

                game.undo()

                //wait before playing so that the user can clearly see the computer's move
                wait(300)

                return EvaluatedMove(randomMove, Ongoing(preliminaryEvaluation, fullEvaluation))
            }
        }
        //the library has no more information on this sequence
        useLibrary = false
        return null
    }

    private fun computeNextMove(): EvaluatedMove {
        //the game starts with a static game evaluation that considers many things
        //but in the endgame, when only the king of one side can move than the only important thing is
        //how many possible moves the king has left. So the Evaluation strategy has to be changed.
        pickStaticSpaceEvaluationIfNecessary()

        //      TODO  real Timer
        //		long time = System.currentTimeMillis();

        val sortedEvaluatedMoves = concurrencyStrategy.evaluateMovesBestMoveFirst(game, evaluatingMinMax)

        //		final int calls = EvaluatingAsIsNow.getCallCounter();
        //		final int totalNumberOfMoves = evaluation.totalNumberOfMoves();
        //		long duration = System.currentTimeMillis()-time;
        //		double timePerMove = duration/(double)totalNumberOfMoves;
        //		double timePerCall = duration/(double)calls;
        //		System.out.println( "ComputerPlayer.play()" );
        //		System.out.println( "moves:       "+totalNumberOfMoves );
        //		System.out.println( "staticEval:  "+calls );
        //		System.out.println( "duration:    "+(duration)+"ms" );
        //		System.out.println( "timePerMove: "+(timePerMove)+"ms" );
        //		System.out.println( "timePerCall: "+(timePerCall)+"ms" );
        //		System.out.println();

        return pickNextMoveByEvaluation(sortedEvaluatedMoves)
    }

    /**
     * pick the move to make given an evaluated Set of possible moves.
     * (Taking always the best move would lead to a limited number of openings!)
     *
     * @param sortedEvaluatedMoves (set.first is the best move for the ki, set.last the worst)
     * @return the move the ki will make next
     */
    private fun pickNextMoveByEvaluation(sortedEvaluatedMoves: List<EvaluatedMove>): EvaluatedMove {
        val bestMove: EvaluatedMove = sortedEvaluatedMoves.first()

        if (bestMove.value !is NumericalEvaluation) {
            return bestMove
        }

        fun pickEvaluateMoveBy(move: Move) = sortedEvaluatedMoves.first { it.move.equalsMove(move) }

        val okDistanceToBest = .2

        // the weight lies between (0-1]
        // with bestMove will have a weight of 1
        // and a move that is almost okDistanceToBest apart will have a weight of almost 0
        val moveAndLinearWeight: List<Pair<Move, Double>> = LinkedList<Pair<Move, Double>>().apply {
            val bestFullEvaluation = bestMove.value.fullEvaluation
            for((move, evaluation) in sortedEvaluatedMoves) {
                if(evaluation !is NumericalEvaluation) break
                val distanceToBest = bestFullEvaluation-evaluation.fullEvaluation
                if(distanceToBest>=okDistanceToBest) break
                add(Pair(move, (okDistanceToBest-distanceToBest)/okDistanceToBest))
            }
        }
        println("considering: ${moveAndLinearWeight.map { it.first.toString() }.joinToString()}")

        require(moveAndLinearWeight.isNotEmpty())
        if(moveAndLinearWeight.size==1) {
            return pickEvaluateMoveBy(moveAndLinearWeight.first().first)
        }

//        fun printMovesAndWeight(moveAndPercentage: List<Pair<Move, Double>>) {
//            val formatter = DecimalFormat().apply {
//                minimumFractionDigits = 2
//                maximumFractionDigits = 2
//            }
//
//            fun percentageToString(value: Double): String = formatter.format(value)
//            println(moveAndPercentage.joinToString { "${it.first}:${percentageToString(it.second)}" })
//        }
//        printMovesAndWeight(moveAndLinearWeight)

        // make it more than linear probable to pick a better move (the bigger the factor, the more preferable better solutions are)
        val moveAndWeight = moveAndLinearWeight.map { Pair(it.first, Math.pow(it.second, 1.8)) }
        val weightSum = moveAndWeight.map { it.second }.sum()
        // the sum of all percentages will be 1.0 (or close to it because of rounding errors)
        val moveAndPercentage = moveAndWeight.map { Pair(it.first, it.second/weightSum) }

//        fun printMovesAndPercentages(moveAndPercentage: List<Pair<Move, Double>>) {
//            val formatter = DecimalFormat().apply {
//                minimumFractionDigits = 1
//                maximumFractionDigits = 1
//            }
//
//            fun percentageToString(value: Double): String = "${formatter.format(100*value)}%"
//            println(moveAndPercentage.joinToString { "${it.first}:${percentageToString(it.second)}" })
//        }
//        printMovesAndPercentages(moveAndPercentage)


        var randomValueInbetween0and1 = randomNumberGenerator.nextDouble()
        val moveAndPercentageIter = moveAndPercentage.iterator()
        var moveWithPercentage = moveAndPercentageIter.next()
        while(moveAndPercentageIter.hasNext() && randomValueInbetween0and1>moveWithPercentage.second) {
            randomValueInbetween0and1 -= moveWithPercentage.second
            moveWithPercentage = moveAndPercentageIter.next()
        }

        return pickEvaluateMoveBy(moveWithPercentage.first)
    }

    /**
     * the game starts with a static game evaluation that considers many things
     * but in the endgame, when only the king of one side can move than the
     * only important thing is how many possible moves the king has left.
     * So the Evaluation strategy has to be changed.
     */
    private fun pickStaticSpaceEvaluationIfNecessary() {
        // TODO implement better endgame engine
//        if (usesStandardEvaluation && EvaluatingSpace.shouldUseStaticSpaceEvaluation(game)) {
//            // once per game
//            evaluatingMinMax.strategy = EvaluatingSpace()
//            evaluatingMinMax.pruner = FullMovePruner(3, 3, 3)
//            usesStandardEvaluation = false
//        }
    }

    override fun askForPawnPromotionType(pawnPosition: Position) = PawnPromotion.QUEEN

    override fun gameStarts() {
        reset()
        ui.reset()
    }

    override fun gameEnds(endoption: MoveResult, lastMoveByWhite: Boolean) {
        ui.showThoughts(false)
        when (endoption) {
            MoveResult.DRAW -> ui.setBubbleText("draw")
            MoveResult.STALEMATE -> ui.setBubbleText("stalemate")
            MoveResult.CHECKMATE -> {
                val didIWin = lastMoveByWhite == isWhite
                if (didIWin) {
                    ui.setBubbleText("checkmate")
                    ui.setThumb(Thumb.DOWN)
                } else {
                    ui.setBubbleText("good game")
                    ui.setThumb(Thumb.UP)
                }
            }
            MoveResult.THREE_TIMES_SAME_POSITION -> ui.setBubbleText("draw because of\n3x repetition")
            MoveResult.FIFTY_MOVES_NO_HIT -> ui.setBubbleText("draw because of\n50-move rule")
            MoveResult.RESIGN -> ui.setBubbleText("you're already\nresigning ?!")
            else -> ui.setBubbleText(endoption.toString())
        }
    }

    override fun gaveCheck() {
        ui.setBubbleText("check")
    }

    override fun setColor(isWhite: Boolean) {
        this.isWhite = isWhite
    }

    // initializes the default EvaluationStrategy
    private fun reset() {
        evaluatingMinMax.strategy = standardEvaluation
        evaluatingMinMax.pruner = standardPruner
        usesStandardEvaluation = true
        //use the library only if the figures are used in the classical way (no Chess960)
        useLibrary = game.isStandardGame
    }

    fun setSearchTreePruner(pruner: SearchTreePruner) {
        standardPruner = pruner
        evaluatingMinMax.pruner = pruner
    }

    fun setNumberOfCoresToUse(numberOfCoresToUse: Int) {
        concurrencyStrategy = getConcurrencyStrategy(ui::setProgress, numberOfCoresToUse)
    }
}
